feat(plugin-rows): new plugin reorder buttons

hejtful opened the pull request:
[Linear](https://linear.app/serlo/project/make-moving-the-plugins-on-canvas-easier-24b0fb77e755/issues)

[Demo](https://frontend-git-feat-rows-new-reorder-buttons-serlo.vercel.app/entity/repository/add-revision/277232)

hejtful committed:
feat(plugin-rows): replace drag handle with reorder buttons

hejtful committed:
feat(plugin-rows): scroll down-moved-plugin back into view

hejtful committed:
Merge branch 'staging' into feat/rows-new-reorder-buttons

hejtful committed:
feat(editor): reorder plugins animation

hejtful committed:
feat(rows-editor): reorder buttons design

hejtful committed:
refactor(rows-editor): try lazy loading motion animations

LarsTheGlidingSquirrel commented:
The animation looks really nice and helps a lot with keeping track of where stuff is flying! :green_heart:  
Only when the moved element moves off screen, the content jumps and it is a little disorienting to me.

hejtful commented:
Any ideas on how to improve this case? :)

LarsTheGlidingSquirrel commented on packages/editor/src/plugins/rows/components/rows-inner-editor.tsx:
```diff
@@ -16,6 +17,9 @@ import {
   PluginMenuContext,
 } from '../contexts/plugin-menu'
 
+const loadMotionFeatures = () =>
+  import('motion/react').then((res) => res.domMax)
+
```
What was the reason for doing it this way?

hejtful replied:
Reducing the initial load. Before lazy loading, the increase in bundle size was much bigger.

I followed the guide from Motion themselves, just omitted the extra file they recommend, as I didn't think we needed that: https://motion.dev/docs/react-lazy-motion

LarsTheGlidingSquirrel commented on packages/editor/src/plugins/rows/editor-row-renderer.tsx:
```diff
@@ -191,15 +190,15 @@ export function EditorRowRenderer({
-          '[&:has(.rows-editor-renderer-container:focus-within)>.rows-tools]:opacity-0',
-          '[&:hover>.rows-tools]:!opacity-100',
+          // Row controls (left sidebar) styles
+          '[&:focus-within>.row-controls]:opacity-100',
+          '[&:has(.rows-editor-renderer-container:focus-within)>.row-controls]:opacity-0',
+          '[&:hover>.row-controls]:!opacity-100',
           isMultimediaPlugin &&
-            '[&>.rows-tools]:!-left-1 [&>.rows-tools]:!-top-9'
+            '[&>.row-controls]:!-left-1 [&>.row-controls]:!-top-9'
         )}
```
Could these styles also be part of component `RowControls`?

hejtful replied:
This Tailwind way of applying CSS hacks is consistently used by Botho across the project, and I honestly prefer it to passing props along components for CSS hacks.

Ideally, we would try to separate UI components from functional components fully, and then we could re-think this approach. But for now, I'd keep this here.

Please note that I only renamed the classes to follow the new component name.

LarsTheGlidingSquirrel commented on packages/editor/src/plugins/rows/components/row-editor.tsx:
```diff
@@ -36,8 +37,9 @@ export function RowEditor({
-      key={row.id}
+    <motion.div
+      layout
+      transition={{ duration: 0.3, type: 'linear' }}
```
Maybe ask Gregor for transition type and duration in terms of good UX.

hejtful replied:
Gregor saw the demo while the PR was a draft already 😄 

And the desired animation is shown in [the ticket](https://linear.app/serlo/issue/EDTR-169/moving-animation-when-switching-with-arrows) already. Would you say the end result doesn't match the video from the ticket?

LarsTheGlidingSquirrel commented on packages/editor/src/plugins/rows/components/row-controls.tsx:
```diff
@@ -0,0 +1,65 @@
+
+export function RowControls({ rows, index }: RowDragButtonProps) {
+  function handleUp() {
+    rows.move(index, index - 1)
```
Good to know that this exists already. :slightly_smiling_face:

LarsTheGlidingSquirrel reviewed (approved):
Added some comments but nothing blocking merging into staging. Thank you! :green_heart:

LarsTheGlidingSquirrel commented:
> Any ideas on how to improve this case? :)

Hm, either no automatic scrolling or smooth scrolling. No scrolling might move the element out of sight and if you want to move it further you have to find it again first. Without thinking about it much, I think smooth scrolling after a short delay could help.

hejtful commented:
> Hm, either no automatic scrolling or smooth scrolling. No scrolling might move the element out of sight and if you want to move it further you have to find it again first. Without thinking about it much, I think smooth scrolling after a short delay could help.

Thanks for your input! I'll talk about it with Gregor.

The problem is that, when an element is moved above the viewport, the browser scrolls up automatically, and we can't control this jump, or at least I didn't find a way to control it. If Gregor agrees to apply a smooth scroll, I'd need a brainstorm with Botho and you to figure out how to prevent or control this scroll-up behavior 😄

hejtful committed:
refactor(rows-editor): scroll into view before switching animation

hejtful committed:
refactor(rows-editor): add offset when scrolling into view

hejtful commented:
Added smooth scrolling, to the place where the moved plugin will move, right before moving the plugin. On Gregor's request, also added some offset between the plugin and the edge of the window.

LarsTheGlidingSquirrel commented:
Thank you for explaining! :green_heart:

elbotho commented:
really nice with the scrolling 👍

As always there is some nesting fun. It's not great but also no real blocker I guess.
<img width="130" alt="image" src="https://github.com/user-attachments/assets/18495d98-2264-4af5-9f3b-bb05ae23a645">
<img width="126" alt="image" src="https://github.com/user-attachments/assets/ed2c0f91-e88a-4a1e-98ef-9c95d253aeae">
<img width="117" alt="image" src="https://github.com/user-attachments/assets/5ba20a56-9abe-4732-a672-47cbc1089ffe">
